src/libs/decoders/mp3_seek_table.cpp

```diff
@@ -0,0 +1,339 @@
+/**
```

Comment:
Thank you for writing up an explanation in this file - it is very helpful. A question: where is this local file stored - in a working dir? What if user does not have write access to the directory? Is generation of seek table fast enough to be performed on every start of DOSBox, or is it rather unpractical?

Reply:
Yes, it's is a local file in the current-working-directory (and the table is appended with however many mp3's are encountered).

For most users their CWD is the game's "root" directory, and they launch `dosbox <start.bat>` or `dosbox game.exe`, and the seektable is created in the game's directory (only holding tables for that game's MP3s). But let's say they launch DOSBox outside in some parent like EXO and then `cd my_game` before launching `start.bat`. Their seek table file will eventually be much larger and have tables for all their games that use imgmount'ed MP3s. 

Good point regarding the write-protected directory.  Yes, I believe the write would fail; I've now improved it to only write if the ofstream was created properly:

``` c
+    // Update our lookup table file with the new seek points and pcm_frame_count.
+    // Note: the serializer elegantly handles C++ STL objects and is endian-safe.
+    seek_points_table[stream_hash] = seek_points_vector;
+    pcm_frame_count_table[stream_hash] = pcm_frame_count;
+    ofstream outfile(filename, ios_base::trunc | ios_base::binary);
+
+    // Caching our seek table to file is optional.  If the user is blocked due to
+    // security or write-access issues, then this write-phase is skipped. In this
+    // scenario the seek table will be generated on-the-fly on every start of DOSBox.
+    if (outfile.is_open()) {
+        Archive<ofstream> serialize(outfile);
+        serialize << SEEK_TABLE_IDENTIFIER << seek_points_table << pcm_frame_count_table;
+        outfile.close();
+    }
```

This falls backs to generating the seek table on every start of DOSBox.  This is not ideal, but is doable, and the "hit" results in maxing out disk IO and/or CPU until all the MP3s are processed.  Even on a raspberry pi, reading ~60MB worth of MP3s takes about 3 seconds using an SD card capable of 20MB/s. CPU usage will be maxed out for that brief period in time.

Users on big desktop machines won't notice that their entire ripped CD of MP3s are being read.. and surely it can be done in less than 1 second. But none the less, it's not great but it's doable - even for handhelds and SOCs.

That said - I don't think we need to worry it about. There are very few games that use CDROM's that also permit running from write-protected media. Almost all of them expect to write various config and state info into their directories. I have wrestled with this a lot with my squashFS approach and have a patch that lets DOSBox trick games into thinking they can write to a write-protected file. (Some games expect write access to pure data files that they infact never write to).  Here's that patch - https://gitlab.com/krcroft/rompy/blob/master/pc/src/dosbox-svn-allow_readonly.patch

Reply:
In an early versions of this code I wrote the seek table into the user's home directory. But figuring out that path in a cross-platform way (OSX, Windows 7 & 10, Linux) was very messy.  Plus, it then disconnected that data from the game itself - so if the user copied the game to another machine, they would unnecessarily have to creating new tables (which would be slow on a Pi or other SOC). 

Another alternative is adding a config path to dosbox.conf, but that touches more common DOSBox code which I (over the years) and I have tried to minimize that fearing the more I touch, the lower the odds of merging upstream.  When I originally posted the patch, I wanted someone like Qbix to review the concept and say "hey, I like the feature, let's integrate it into DOSBox by doing A, B, and C".  So that's where I'm hoping he would weigh in on details like this.  But frankly, I think it makes a lot of sense to carry the file along inside the game directory itself (given it's cross-platform and endian-safe).

Reply:
I asked this question, because through work on Boxtron I saw, that some game publishers do pretty ridiculous setups when distributing DOS games, e.g. via Steam - in the most degenerate example I saw working directory set to A, game installed in B imgmounting a broken .cue file located in C, that had relative paths to files stored in directory D (Carmageddon distributed via Steam - but it uses .ogg files), so now I am sensitive to where the working dir might be ;). Also, DOSBox installed via Flatpak or Snap might limit write access to some directories (default configuration of DOSBox flatpak package is very strict about it).

However, simply using working dir should be fine in > 99% of cases - it might cause us some issues when we'll work on rom-like package files for DOS games in the future, but if a file can be included inside a package, then it should work great.

As for figuring out app-specific user config paths - SDL2 has this functionality built-in (not only for Linux, Windows and macOS, but also e.g. Android and iOS), so once SDL2 will be implemented, we can use that.

Reply:
Fixed in https://github.com/dreamer/dosbox-staging/commit/c4b303a2414482fa5a4a02d3c094e83fba32dcf4